Donation History UI is now polished

henryholtgeerts opened the pull request:
Resolves #5562 

## Description
This PR introduces style improvements to the Donor History UI components (visible in donation tables and donation receipt pages of the Donor Profile).

## Affects
This PR primarily affects frontend styling and some backend logic related to displaying donation history in Donor Profiles.

## Visuals

<img width="653" alt="Screen Shot 2021-01-25 at 5 32 24 PM" src="https://user-images.githubusercontent.com/5186078/105774508-5640f300-5f33-11eb-96e8-e7e147dfe52b.png">

<img width="651" alt="Screen Shot 2021-01-25 at 5 32 09 PM" src="https://user-images.githubusercontent.com/5186078/105774531-593be380-5f33-11eb-92f2-998f3cf634d3.png">

## Pre-review Checklist

-   [x] Acceptance criteria satisfied and marked in related issue
-   [x] Relevant `@since` tags included in DocBlocks
-   [x] Changes logged to the `Unreleased` section of `CHANGELOG.md`
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

## Testing Instructions

1. Add a Donor Profile block to a new page
2. On the frontend, check that the donation history displayed in Donor Profile matches currently logged in Donor's donaiton history
3. Check that UI components match Figma designs (https://www.figma.com/file/BKr2K0XdDyRRoccJbJKRD7/Epic%3A-Donor-Profile?node-id=397%3A1743)

DevinWalker commented:
@henryholtgeerts @JasonTheAdams @ravinderk what's the status of this PR?

ravinderk commented:
@DevinWalker @JasonTheAdams This was not on my radar. I will review this today.

ravinderk commented on src/DonorProfiles/resources/js/app/components/donation-receipt/index.js:
```diff
@@ -44,7 +44,9 @@ const DonationReceipt = ( { donation } ) => {
 						</div>
 						<div className="give-donor-profile-donation-receipt__value">
-							{ donor.address }
+							{ donor.address.street } <br />
```
@henryholtgeerts can we make address formatting dynamic because an extra line will be added if any entry is missing.

henryholtgeerts replied:
Good thinking! Done, ref: https://github.com/impress-org/givewp/pull/5566/commits/5cccf424904c6d3a39f2a372257eb2584ecc8f4c

ravinderk commented on src/DonorProfiles/resources/js/app/components/donation-table/donation-row/index.js:
```diff
@@ -14,18 +15,27 @@ const DonationRow = ( { donation } ) => {
 				</div>
+				{ payment.mode !== 'live' && (
+					<div className="give-donor-profile-table__donation-test-tag">
+						{ __( 'Test Donation', 'givewp' ) }
```
@henryholtgeerts text-domain is `give`. Can you fix the typo?

henryholtgeerts replied:
Done, ref: https://github.com/impress-org/givewp/pull/5566/commits/1e5d1a5ce58ef7b9d3ade6fe4766116b3cd85492

ravinderk commented on src/DonorProfiles/Repositories/Donations.php:
```diff
@@ -193,15 +193,65 @@ protected function getFormInfo( $payment ) {
+			'total'    => $this->getFormattedAmount( $payment->total, $payment ),
+			'method'   => $gateways[ $payment->gateway ]['checkout_label'],
+			'status'   => $this->getFormattedStatus( $payment->status ),
+			'date'     => date( 'F j, Y', strtotime( $payment->date ) ),
```
@henryholtgeerts Is the time format hardcoded on the donor profile page? in the core, we show the time format as per the website setting.

![image](https://user-images.githubusercontent.com/1784821/107025109-f4a04480-67ce-11eb-8e8c-d0f865dd7b2c.png)

henryholtgeerts replied:
Ah - good suggestion! I was working off the Figma designs, but this is important to consider. Done, ref: https://github.com/impress-org/givewp/pull/5566/commits/782f2a24c783dd7dc8b76cc848dc6e1d8ddfc008

ravinderk reviewed (changes requested):
@henryholtgeerts  I noticed few bugs. May be there happeing because reactjs or PHP error. Can you fix them if any?

1. I am getting a fatal error on PHP 5.6
![image](https://user-images.githubusercontent.com/1784821/106992411-6744fc00-679e-11eb-9ec4-952d85886328.png)
2. I am seeing a warning on the donor profile page:
![image](https://user-images.githubusercontent.com/1784821/106992926-927c1b00-679f-11eb-8bf6-659f432da830.png)

To fix this issue, pass `true` as second param and function will return single value instead of `array`.
![image](https://user-images.githubusercontent.com/1784821/106993137-133b1700-67a0-11eb-9747-15d6985c99ab.png)
3. I am getting react warning on the donor profile page:
![image](https://user-images.githubusercontent.com/1784821/106993278-61e8b100-67a0-11eb-9f8a-922518c67e6b.png)
4. Is the profile page UI fully responsive? what device do we support?
![image](https://user-images.githubusercontent.com/1784821/106993394-9d837b00-67a0-11eb-9246-7a79576645f1.png)
5. Can we set pointer cursor icon to pagination links:
![image](https://user-images.githubusercontent.com/1784821/107000605-26ed7a00-67ae-11eb-876d-4965a7f207e0.png)
6. I am input field outline, is it intentional:
![image](https://user-images.githubusercontent.com/1784821/107024142-a9396680-67cd-11eb-9419-359510894d55.png)
7. Arrow icon is missing from links:
![image](https://user-images.githubusercontent.com/1784821/107024259-d1c16080-67cd-11eb-91ed-c9e3d507b4f7.png)
8. Print icon is missing and few icon does not have color on donor profile page:
![image](https://user-images.githubusercontent.com/1784821/107024676-58763d80-67ce-11eb-872e-e3023d6d28c1.png)
9. I am not seeing donor information correctly:
![image](https://user-images.githubusercontent.com/1784821/107025398-65dff780-67cf-11eb-9f6d-c50a38eea068.png)
![image](https://user-images.githubusercontent.com/1784821/107025451-785a3100-67cf-11eb-8202-458ee98a1e89.png)
10. I am not see login form for on loggedin donor:
![image](https://user-images.githubusercontent.com/1784821/107025603-ad668380-67cf-11eb-8315-6afbba92b0c2.png)
![image](https://user-images.githubusercontent.com/1784821/107025650-bce5cc80-67cf-11eb-8097-4f8f658628c1.png)
11. Are addon tabs hardcoded? I can see recurring donation tab if recurring donations addon disbaled. Admin can deactivate recurring donation if all subscription completed. We have to make sure which action we can show to dnor if recurring donation deactivate.

![image](https://user-images.githubusercontent.com/1784821/107025960-18b05580-67d0-11eb-870e-b586a705465c.png)
12. I am see unnecessary text on pagination links location:
![image](https://user-images.githubusercontent.com/1784821/107026258-76dd3880-67d0-11eb-886a-ce12664f91be.png)
13. can donor exit `Prefix` even if `Name Title Prefix` setting disabled :
![image](https://user-images.githubusercontent.com/1784821/107026411-ad1ab800-67d0-11eb-8598-2a69de5bd302.png)
14. Where i can see subscription initial payment and renewal. Do we have any UI to show them different or mention that these payments are for subscription:
![image](https://user-images.githubusercontent.com/1784821/107026600-f2d78080-67d0-11eb-8d76-ed49fbb79916.png)
15. Active button color does not match design. text also not align with button as in design:
![image](https://user-images.githubusercontent.com/1784821/107026777-39c57600-67d1-11eb-82ac-06ab09af5a98.png)
16. Do we show banner on dashboard?
![image](https://user-images.githubusercontent.com/1784821/107026890-65e0f700-67d1-11eb-9d0b-e12a2bb8caee.png)
17. I am not seeing `Logout` link:
![image](https://user-images.githubusercontent.com/1784821/107027082-a93b6580-67d1-11eb-8346-c0994281ce27.png)

henryholtgeerts commented on src/DonorProfiles/Repositories/Donations.php:
```diff
@@ -193,15 +193,65 @@ protected function getFormInfo( $payment ) {
+		$statusMap = [
+			'publish' => [
+				'color' => '#7AD03A',
+				'label' => __( 'Complete', 'give' ),
```
Done, ref: https://github.com/impress-org/givewp/pull/5566/commits/54a2ca0f735e77f97ed0f1ecf50b6324d8a95778

henryholtgeerts commented:
@ravinderk Thanks for the thorough review! I've addressed your specific requested changes and resolved those comments. I'll address the changes you suggested in the larger comment here:

1. Good catch, I've removed the type declaration. Done, ref: https://github.com/impress-org/givewp/pull/5566/commits/7bd23cab909ea19daabac4f257872c99e38d4d6d

2. Thanks for the tip! Done, ref: https://github.com/impress-org/givewp/pull/5566/commits/00523a5070d61b116587a4ff1ba44ae4e7bb5cbe

3. I've now added a key prop for Donation Rows to prevent this issue. Done, ref: https://github.com/impress-org/givewp/pull/5566/commits/f47916ddcc3e9f5c51886f9bf2cec78e94791919

4. Making the profile page UI responsive will be handled in a future PR.

5. Good idea! Done, ref: https://github.com/impress-org/givewp/pull/5566/commits/7d008e908000dfa8d57b507fe86f64a715f31815

6. The inner outline you point out is added by theme styles. This will be resolved in a future PR which will encapsulate Donor Profile styles via the shadow DOM.

7. Recurring Donation UI is being handled in a seperate PR: https://github.com/impress-org/givewp/issues/5582

8. PDF Receipts and Donor Badges are being handled in a future PR. 

9. This is interesting -- was the Donor Profile updated, then Donation History was viewed? This might explain why new Donation history was not retrieved to display the correct Donor Information. Currently, Donation History is only retrieved once, on page load.

10. Login UI is being handled in a separate PR: https://github.com/impress-org/givewp/pull/5569

11. Recurring Donation UI is being handled in a seperate PR: https://github.com/impress-org/givewp/issues/5582

12. Recurring Donation UI is being handled in a seperate PR: https://github.com/impress-org/givewp/issues/5582

13. Edit Profile UI was handled in a previous PR, at that time it was decided that the Title prefix option should be made available to donors regardless of global settings: https://github.com/impress-org/givewp/pull/5463

14. Recurring Donation UI is being handled in a seperate PR: https://github.com/impress-org/givewp/issues/5582

15. In a previous PR, it was determined that the default accent color for the Donor Profile app should be GiveWP green, not the blue as depicted in the Figma designs. Ultimately, admin will be able to set the accent color manually.

16. The Give Again banner will be addressed in a future PR.

17. Logout UI is being handled in a separate PR: https://github.com/impress-org/givewp/pull/5569

ravinderk commented:
@henryholtgeerts I find few issues. Can you review them?


- [ ] Can your review all code related to DonorProfile because I am still seeing `PHP 5.6` incompatible code
![image](https://user-images.githubusercontent.com/1784821/107790274-9ba95100-6d78-11eb-9042-dee3da3c6d6c.png)

- [ ] I am seeing a javascript error in the console tab.
![image](https://user-images.githubusercontent.com/1784821/107790588-f2168f80-6d78-11eb-887e-3a6c33caef6e.png)

- [ ] I am getting PHP notices when loading the donor profile page.
```
[12-Feb-2021 15:57:01 UTC] PHP Notice:  Undefined index: give_subscription in /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/DonorProfiles/Repositories/Donations.php on line 227
[12-Feb-2021 15:57:01 UTC] PHP Stack trace:
[12-Feb-2021 15:57:01 UTC] PHP   1. {main}() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:0
[12-Feb-2021 15:57:01 UTC] PHP   2. require() /Users/ravinderkumar/Local-Sites/give/app/public/index.php:17
[12-Feb-2021 15:57:01 UTC] PHP   3. wp() /Users/ravinderkumar/Local-Sites/give/app/public/wp-blog-header.php:16
[12-Feb-2021 15:57:01 UTC] PHP   4. WP->main() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/functions.php:1291
[12-Feb-2021 15:57:01 UTC] PHP   5. WP->parse_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:745
[12-Feb-2021 15:57:01 UTC] PHP   6. do_action_ref_array() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp.php:388
[12-Feb-2021 15:57:01 UTC] PHP   7. WP_Hook->do_action() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/plugin.php:551
[12-Feb-2021 15:57:01 UTC] PHP   8. WP_Hook->apply_filters() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:311
[12-Feb-2021 15:57:01 UTC] PHP   9. rest_api_loaded() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/class-wp-hook.php:287
[12-Feb-2021 15:57:01 UTC] PHP  10. WP_REST_Server->serve_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api.php:354
[12-Feb-2021 15:57:01 UTC] PHP  11. WP_REST_Server->dispatch() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:440
[12-Feb-2021 15:57:01 UTC] PHP  12. WP_REST_Server->respond_to_request() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1007
[12-Feb-2021 15:57:01 UTC] PHP  13. Give\DonorProfiles\Routes\DonationsRoute->handleRequest() /Users/ravinderkumar/Local-Sites/give/app/public/wp-includes/rest-api/class-wp-rest-server.php:1160
[12-Feb-2021 15:57:01 UTC] PHP  14. Give\DonorProfiles\Routes\DonationsRoute->getData() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/DonorProfiles/Routes/DonationsRoute.php:45
[12-Feb-2021 15:57:01 UTC] PHP  15. Give\DonorProfiles\Repositories\Donations->getDonations() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/DonorProfiles/Routes/DonationsRoute.php:78
[12-Feb-2021 15:57:01 UTC] PHP  16. Give\DonorProfiles\Repositories\Donations->getPaymentInfo() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/DonorProfiles/Repositories/Donations.php:167
[12-Feb-2021 15:57:01 UTC] PHP  17. Give\DonorProfiles\Repositories\Donations->getFormattedStatus() /Users/ravinderkumar/Local-Sites/give/app/public/wp-content/plugins/give/src/DonorProfiles/Repositories/Donations.php:205
```
- [ ] Text and icons are not aligned
![image](https://user-images.githubusercontent.com/1784821/107791283-c0ea8f00-6d79-11eb-82b5-e60c247fe523.png)

henryholtgeerts commented:
@ravinderk Thanks for the additional feedback! I've addressed each items in these commits:

1. https://github.com/impress-org/givewp/pull/5566/commits/875a2e89fb8008baf447b54973a9ff714a05b95d
2. https://github.com/impress-org/givewp/pull/5566/commits/90c74378658694635352c30ca2ed23e252281292
3. https://github.com/impress-org/givewp/pull/5566/commits/2a01060c7a4b79b8ddaaa3978bd82d42add380d2
4. https://github.com/impress-org/givewp/pull/5566/commits/aefae6782f3407e49155876b047b5c29ef21a23e
